src/lib/pubkey/rfc6979/rfc6979.cpp

```diff
@@ -27,7 +27,15 @@ RFC6979_Nonce_Generator::RFC6979_Nonce_Generator(std::string_view hash, const Bi
 RFC6979_Nonce_Generator::~RFC6979_Nonce_Generator() = default;
 
 const BigInt& RFC6979_Nonce_Generator::nonce_for(const BigInt& m) {
-   m.serialize_to(std::span{m_rng_in}.subspan(m_rlen));
+   std::vector<uint8_t> m_bytes(m_rlen);
+   m.serialize_to(m_bytes);
+   return this->nonce_for(m_bytes);
+}
```

Comment:
_Not sure its worth omitting the extra allocation,_ but you could serialize straight into `m_rng_in` and then establish an internal method that just assumes `m_rng_in` is properly set up. Similarly, for the `std::span<>` overload, that could just `copy_mem` into `m_rng_in` and then also call this private method.

Like so:

```C++
m.serialize_to(std::span{m_rng_in}.subspan(m_rlen));
return this->generate_nonce();
```

While we're on it: it might be worthwhile to keep the span "`std::span{m_rng_in}.subspan(m_rlen)`" as a member variable to avoid misuse (by malforming the `subspan()` invocation) here and in the other overload

Reply:
Good idea. I think the original version of this was written before we finalized on the write-to-span idiom for BigInt.

Reply:
Actually looked at this more and it doesn't seem worth bothering with.

There are two uses of RFC 6979 right now, ECDSA and DSA.

* ECDSA uses the bytes variant (or will, with this PR merged)
* DSA actually sets up a complete new `RFC6979_Nonce_Generator` object for each signature! Saving one extra allocation is the least of our worries :see_no_evil: 

Any future users of RFC 6979 (eg ECGDSA or etc) will use the bytes interface, so this second overload of `nonce_for` really is (will become) DSA specific at this point.

I think this can be cleaned up further - in particular in the long run I'd like to avoid the multiple conversions between `BigInt` and `EC_Scalar` implied by

```
   const auto k = EC_Scalar::from_bigint(m_group, m_rfc6979->nonce_for(m_bytes: m.serialize()));
```

It's interesting that as code becomes faster, tiny optimizations become more meaningful. We're already under 250K cycles for ECDSA signatures for secp256r1 (for pcurves). At this point saving even a thousand cycles can be noticeable in the overall throughput.

I left a todo regarding this in #4027

Reply:
I was somewhat surprised in #4143 that ECDSA once bridged was slower than our "demo" ECDSA directly using pcurves. I initially ascribed it to overhead of the abstractions but it turns out almost all of the overhead is RFC 6979. Almost 50K cycles on my laptop. I think, while implementing this in terms of `HMAC_DRBG` was convenient, it's not very fast...